### PR TITLE
Versioning Changes

### PR DESCRIPTION
**Pull request by brthor**
Changes include:
1.Better command line option parsing by the package tool. This allows versions and names to be overridden.
2.Support for CONFLICTS field in debian control. This allows for declaration of incompatible packages.
3.Up to date documentation
4.Changes to build scripts to take advantage of these changes.
5.Change the build version variable from  DOTNET_BUILD_VERSION  to  DOTNET_CLI_VERSION  
6.Implement new versioning scheme.

### CODE REVIEW

File: scripts/package/package-debian.sh [hunk 1]
```diff
@@ -32,12 +32,28 @@ TEST_STAGE_DIR="$PACKAGE_OUTPUT_DIR/test"
...
+    determine_package_name
     create_empty_debian_layout
     copy_files_to_debian_layout
     create_debian_package
 }
 
+determine_package_name(){
+    if [[ "$RELEASE_SUFFIX" == "dev" ]]; then
+        DOTNET_DEB_PACKAGE_NAME=$NIGHTLY_PACKAGE_NAME
+    elif [[ "beta rc1 rc2 rtm" =~ (^| )"$RELEASE_SUFFIX"($| ) ]]; then
```
Comment (brthor) [inline scripts/package/package-debian.sh]: @piotrpMSFT @Sridhar-MS Are there any more identifiers we may want to consider here?

Comment (Sridhar-MS) [inline scripts/package/package-debian.sh]: This looks ok for now. We add more to the list in case we need it later.

**Comment by brthor** [comment]
Caused by: java.io.IOException: remote file operation failed: /mnt/j/workspace/dotnet_cli/debug_ubuntu_prtest at hudson.remoting.Channel@6f4b8df5:dci-ub-bld-12: java.io.IOException: No space left on device

Noooooo :frowning:

**Comment by brthor** [comment]
@dotnet-bot test Ubuntu Debug Build please

**Comment by brthor** [comment]
@dotnet-bot test Ubuntu Release Build please

File: scripts/build/generate-version.ps1 [hunk 2]
```diff
@@ -3,17 +3,14 @@
...
-$buildnumber = 0
-$revnumber = $commitTime.TotalSeconds
+# Zero Padded Suffix for use with Nuget
+$env:VersionSuffix = "$ReleaseSuffix-$CommitCountVersion"
 
-$VersionSuffix = "dev-$revnumber"
-
-$env:DOTNET_BUILD_VERSION = "$majorVersion.$minorVersion.$buildnumber.$revnumber"
\ No newline at end of file
+$env:DOTNET_CLI_VERSION = "$MajorVersion.$MinorVersion.$PatchVersion.$CommitCountVersion"
```
Comment (Sridhar-MS) [inline scripts/build/generate-version.ps1]: It should be `$env:DOTNET_CLI_VERSION = "$env:MajorVersion.$env:MinorVersion.$env:PatchVersion.$env:CommitCountVersion" 
`

Comment (Sridhar-MS) [inline scripts/build/generate-version.ps1]: Let me fix this

Comment (Sridhar-MS) [inline scripts/build/generate-version.ps1]: @brthor I dont have access to fix this in your fork. Can you provide me access or fix this yourself?

File: packaging/nuget/package.ps1 [hunk 3]
```diff
@@ -10,7 +10,7 @@ param(
 $toolsDir = $toolsDir.TrimEnd('\')
 $versionArg = ""
 if ($versionSuffix -ne "") {
-    $versionArg = "--version-suffix $VersionSuffix"
+    $versionArg = "--version-suffix $versionSuffix"
```
Comment (Sridhar-MS) [inline packaging/nuget/package.ps1]: Does it have to be `$env:VersionSuffix`?

Comment (Sridhar-MS) [inline packaging/nuget/package.ps1]: Ignore the above command.

**Comment by Sridhar-MS** [comment]
You may have to make a chnage here to use the env var - https://github.com/dotnet/cli/blob/master/scripts/package/package.ps1#L16

File: packaging/nuget/package.ps1 (hunk 3 above)
Comment (brthor) [inline packaging/nuget/package.ps1]: This one is using the arg above.

File: scripts/build/generate-version.ps1 (hunk 2 above)
Comment (brthor) [inline scripts/build/generate-version.ps1]: Will do both. Thanks!

**Comment by brthor** [comment]
Thanks for pointing that out, will make the change

**Comment by brthor** [comment]
@dotnet-bot test this please

**Pull request merged**